chord_metadata_service/restapi/candig_authz_middleware.py

```diff
@@ -0,0 +1,147 @@
+from django.conf import settings
+from django.http import HttpResponseForbidden
+import requests
+import re
+import json
+
+
+class CandigAuthzMiddleware:
+    """
+    A generic middleware for CanDIGv2 authorization.
+    """
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+        self.authorize_datasets = 'NO_DATASETS_AUTHORIZED'
+
+
+    def __call__(self, request):
+        if settings.CANDIG_AUTHORIZATION == 'OPA':
+            token = self.get_auth_token(request.headers)
+            if self.is_authorized_get(request):
+                """
+                You may incorporate any data source for dataset authorization.
+        
+                Note, if no datasets are authorized, you MUST set it to 'NO_DATASETS_AUTHORIZED'.
+        
+                If authorized datasets exist, set it to a comma-separated string with titles of authorized datasets.
+        
+                For example, if datasets d100 and d200 are authorized, you should set
+                self.authorize_datasets = 'd100,d200'
+                """
+                if settings.CACHE_TIME != 0:
+                    error_response = {
+                        "error": "This request failed because caching is not disabled. Please contact your system administrator for assistance."
+                    }
+                    response = HttpResponseForbidden(json.dumps(error_response))
+                    response["Content-Type"] = "application/json"
+                    return response
+                opa_res_datasets = self.get_opa_datasets(token, request.path, request.method)
+                if len(opa_res_datasets) == 0:
+                    self.authorize_datasets = 'NO_DATASETS_AUTHORIZED'
+                elif type(opa_res_datasets) == tuple and opa_res_datasets[0] == "error":  # error response
+                    return opa_res_datasets[1]
```

Comment:
So OPA sends an error and the next element is the error message or a status code?

Reply:
Yeah, an error message